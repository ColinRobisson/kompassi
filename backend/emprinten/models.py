from __future__ import annotations

import os
import typing

from django.db import models


class Project(models.Model):
    name = models.CharField(max_length=255, null=False)
    slug = models.SlugField(unique=True, null=False)

    split_output = models.BooleanField(
        null=False,
        help_text="If set, produce a zip archive containing one or more PDFs instead of a single PDF.",
    )
    name_pattern = models.CharField(
        max_length=255,
        null=False,
        blank=True,
        help_text="File name template pattern without file extension."
        " Gets row dictionary as context unless rendering multiple rows and Split output is not set.",
    )
    title_pattern = models.CharField(
        max_length=255,
        null=False,
        help_text="PDF title template pattern."
        " Gets row dictionary as context unless rendering multiple rows and Split output is not set.",
    )

    def __str__(self) -> str:
        return f"{self.slug}: {self.name}"

    def current_files(self) -> typing.Iterable[FileVersion]:
        return FileVersion.objects.filter(file__project=self, current=True).select_related("file")


class ProjectFile(models.Model):
    class Type(models.TextChoices):
        Main = "main"
        HTML = "html"
        CSS = "css"
        CSV = "csv"
        Image = "img"

    project = models.ForeignKey(Project, on_delete=models.CASCADE, null=False)
    file_name = models.CharField(max_length=255, null=False, unique=True)
    type = models.CharField(max_length=4, null=False, choices=Type.choices)

    hidden = models.BooleanField(default=False, null=False)
    editable = models.BooleanField(default=False, null=False)
    transient = models.BooleanField(default=False, null=False)

    def __str__(self) -> str:
        return f"{self.project}/{self.file_name} ({self.type})"


def make_filename(instance: FileVersion, filename: str) -> str:
    parts = os.path.splitext(os.path.basename(filename))
    if parts[1]:
        name = f"{parts[0]}-{instance.version}{parts[1]}"
    else:
        name = f"{parts[0]}-{instance.version}"
    return os.path.join(
        instance._meta.app_label,
        instance.file.project.slug,
        name,
    )


class FileVersion(models.Model):
    file = models.ForeignKey(ProjectFile, on_delete=models.CASCADE, null=False)
    data = models.FileField(upload_to=make_filename, null=False)
    version = models.PositiveIntegerField(default=1, null=False)
    current = models.BooleanField(default=False)

    class Meta:
        constraints = [
            models.UniqueConstraint(
                fields=["file"],
                condition=models.Q(current=True),
                name="unique_current_file",
            ),
        ]
        unique_together = (("file", "version"),)

    def __str__(self) -> str:
        return f"{self.data.name} (v{self.version})"
