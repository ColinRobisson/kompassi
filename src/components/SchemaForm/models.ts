export type FieldType = 'SingleLineText' | 'MultiLineText' | 'Divider' | 'StaticText' | 'Spacer';


interface BaseField {
  type: FieldType;
  name: string;
  title?: string;
  helpText?: string;
  required?: boolean;
  readOnly?: boolean;
}


export interface SingleLineText extends BaseField {
  type: 'SingleLineText';
}


export interface MultiLineText extends BaseField {
  type: 'MultiLineText';
  rows?: number;
}


export interface Divider extends BaseField {
  type: 'Divider';
}

export interface Spacer extends BaseField {
  type: 'Spacer';
}


export interface StaticText extends BaseField {
  type: 'StaticText';
}


export interface SingleLineText extends BaseField {
  type: 'SingleLineText';
}

export type Field = SingleLineText | MultiLineText | Divider | Spacer | StaticText;

export type Layout = 'horizontal' | 'vertical'; // TODO: inline
