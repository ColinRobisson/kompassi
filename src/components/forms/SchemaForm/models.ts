export type FieldType = 'SingleLineText' | 'MultiLineText' | 'Divider' | 'StaticText' | 'Spacer' | 'SingleCheckbox';
export const fieldTypes: FieldType[] = ['SingleLineText', 'MultiLineText', 'SingleCheckbox', 'StaticText', 'Divider', 'Spacer'];

interface BaseField {
  type: FieldType;
  name: string;
  title?: string;
  helpText?: string;
  required?: boolean;
  readOnly?: boolean;
}

export interface SingleLineText extends BaseField {
  type: 'SingleLineText';
}

export interface MultiLineText extends BaseField {
  type: 'MultiLineText';
  rows?: number;
}

export interface Divider extends BaseField {
  type: 'Divider';
}

export interface Spacer extends BaseField {
  type: 'Spacer';
}

export interface StaticText extends BaseField {
  type: 'StaticText';
}

export interface SingleLineText extends BaseField {
  type: 'SingleLineText';
}

export interface SingleCheckbox extends BaseField {
  type: 'SingleCheckbox';
}

export type Field = SingleLineText | MultiLineText | Divider | Spacer | StaticText | SingleCheckbox;

export type Layout = 'horizontal' | 'vertical';

export interface Form {
  title: string;
  slug: string;
  fields: Field[];
  layout: Layout;
}
