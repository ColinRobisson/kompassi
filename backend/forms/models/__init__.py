from .dimension import Dimension, DimensionValue, ResponseDimensionValue
from .field import Field, FieldType
from .form import Form
from .keypair import KeyPair
from .meta import FormsEventMeta, FormsProfileMeta
from .response import Response
from .survey import Survey
