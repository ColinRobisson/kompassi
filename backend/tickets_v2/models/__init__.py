from .meta import TicketsV2EventMeta
from .order import Order
from .product import Product
from .quota import Quota
from .ticket import Ticket
