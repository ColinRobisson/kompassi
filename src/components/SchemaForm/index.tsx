import React from 'react';
import { NamespacesConsumer } from 'react-i18next';

import Button from 'reactstrap/lib/Button';
import ButtonGroup from 'reactstrap/lib/ButtonGroup';
import Col from 'reactstrap/lib/Col';
import Form from 'reactstrap/lib/Form';
import FormGroup from 'reactstrap/lib/FormGroup';
import Input from 'reactstrap/lib/Input';
import Label from 'reactstrap/lib/Label';

import { Field, FieldType } from './models';


interface SchemaFormProps {
  fields: Field[];
  layout?: 'vertical' | 'horizontal';
  ns?: string[];
}


interface SchemaFormState { }


export default class SchemaForm extends React.PureComponent<SchemaFormProps, SchemaFormState> {
  static defaultNamespace = 'SchemaForm';

  render() {
    const { layout } = this.props;
    const ns = (this.props.ns || []).concat(['SchemaForm']);

    return (
      <NamespacesConsumer ns={ns}>
        {t => (
          <Form className={layout === 'horizontal' ? 'form-horizontal' : ''}>
            {this.props.fields.map(this.renderField)}

            <ButtonGroup className="float-md-right">
              <Button color="primary">{t('submit')}</Button>
            </ButtonGroup>
          </Form>
        )}
      </NamespacesConsumer>
    );
  }

  renderField = (field: Field) => {
    const { layout } = this.props;
    const { name, title, helpText } = field;

    switch (layout) {
      case 'horizontal':
        return (
          <FormGroup key={name} className="row">
            <Label for={name} className="col-md-3 col-form-label">{title}</Label>
            <Col md={9}>{this.renderInput(field)}</Col>
          </FormGroup>
        );
      case 'vertical':
      default:
        return (
          <FormGroup key={name} className="row">
            <Label for={name}>{title}</Label>
            {this.renderInput(field)}
          </FormGroup>
        );
    }
  }

  renderInput = (field: Field) => {
    switch (field.type) {
      case 'Input':
        return <Input name={field.name} />;
      case 'TextArea':
        return <Input type="textarea" name={field.name} rows={field.rows || 10} />;
    }
  }
}
