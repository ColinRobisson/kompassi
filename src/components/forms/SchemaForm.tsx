import React from "react";

import { useFormik } from "formik";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";
import Spinner from "react-bootstrap/Spinner";

import { FormSchema } from "./models";
import { schemaToYup } from "./validation";
import SchemaFormInput from "./SchemaFormInput";
import SchemaFormField from "./SchemaFormField";

interface SchemaFormProps {
  schema: FormSchema;
  initialValues?: any;
  showSubmitButton?: boolean; // default true
  onSubmit(values: any): void;
}

const SchemaForm = ({
  schema,
  initialValues,
  onSubmit,
  showSubmitButton,
}: SchemaFormProps) => {
  initialValues = initialValues ?? {};

  const validationSchema = React.useMemo(() => schemaToYup(schema), [schema]);
  const { layout } = schema;
  const { handleSubmit, handleChange, values, errors, isSubmitting } =
    useFormik({
      initialValues,
      onSubmit,
      validationSchema,
    });

  return (
    <Form noValidate onSubmit={handleSubmit}>
      {schema.fields.map((field) => (
        <SchemaFormField
          layout={layout}
          key={field.name}
          field={field}
          error={errors[field.name]}
        >
          <SchemaFormInput
            field={field}
            value={values[field.name] || ""}
            onChange={handleChange}
            error={errors[field.name]}
          />
        </SchemaFormField>
      ))}
      <Button
        type="submit"
        disabled={isSubmitting}
        className={showSubmitButton ?? true ? "" : "sr-only"}
      >
        {(isSubmitting && (
          <>
            <Spinner animation="border" size="sm" /> Submitting...
          </>
        )) ||
          "Submit"}
      </Button>
    </Form>
  );
};
export default SchemaForm;
