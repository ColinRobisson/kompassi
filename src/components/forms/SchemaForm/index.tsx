import React, { FormEvent, ChangeEvent } from 'react';

import Col from 'reactstrap/lib/Col';
import Form from 'reactstrap/lib/Form';
import FormGroup from 'reactstrap/lib/FormGroup';
import FormText from 'reactstrap/lib/FormText';
import Input from 'reactstrap/lib/Input';
import Label from 'reactstrap/lib/Label';

import { Field, Layout } from './models';
import HorizontalField from '../HorizontalField';

interface SchemaFormProps {
  fields: Field[];
  layout?: Layout;
  ns?: string[];

  // TODO Stricter typings
  /* eslint-disable @typescript-eslint/no-explicit-any */
  value?: any;
  onChange?(values: any): void;
  onSubmit?(values: any): void;
  /* eslint-enable @typescript-eslint/no-explicit-any */
}

export class BaseSchemaForm<OwnProps> extends React.PureComponent<SchemaFormProps & OwnProps, {}> {
  render() {
    const { layout, fields, children } = this.props;

    return (
      <Form className={layout === 'horizontal' ? 'form-horizontal' : ''}>
        {fields.map(field => (
          <div key={field.name}>{this.renderField(field)}</div>
        ))}
        {children}
      </Form>
    );
  }

  protected renderField(field: Field) {
    const { layout, value } = this.props;
    const { type, name, helpText, required } = field;
    const title = field.title || '';
    const fieldValue = value ? value[field.name] || '' : '';

    if (type === 'StaticText' && !title) {
      // Full-width static text
      return <p>{helpText}</p>;
    } else if (type === 'Divider') {
      return <hr />;
    } else if (type === 'Spacer') {
      return <div className="pb-3" />;
    }

    switch (type) {
      case 'SingleCheckbox':
        return (
          <FormGroup check={true}>
            <Label check={true} className={required ? 'font-weight-bold' : ''}>
              {this.renderInput(field, fieldValue)} {title}
              <FormText>{helpText}</FormText>
            </Label>
          </FormGroup>
        );
      default:
        switch (layout) {
          case 'horizontal':
            return (
              <HorizontalField name={name} title={title} helpText={helpText} required={required}>
                {this.renderInput(field, fieldValue)}
              </HorizontalField>
            );
          case 'vertical':
          default:
            return (
              <FormGroup>
                <Label for={name} className={required ? 'font-weight-bold' : ''}>
                  {title}
                </Label>
                {this.renderInput(field, fieldValue)}
                <FormText>{helpText}</FormText>
              </FormGroup>
            );
        }
    }
  }

  protected renderInput(field: Field, value: any) {
    const readOnly = this.isReadOnly(field);

    switch (field.type) {
      case 'SingleLineText':
        return <Input id={field.name} name={field.name} readOnly={readOnly} onChange={this.handleTextFieldChange} value={value} />;
      case 'MultiLineText':
        return (
          <Input
            type="textarea"
            id={field.name}
            name={field.name}
            readOnly={readOnly}
            onChange={this.handleTextFieldChange}
            rows={field.rows || 10}
            value={value}
          />
        );
      case 'SingleCheckbox':
        return (
          <Input
            type="checkbox"
            id={field.name}
            name={field.name}
            disabled={readOnly}
            onChange={this.handleCheckboxChange}
            value={value}
          />
        );
      case 'StaticText':
        return <></>;
      default:
        throw new Error(`Not implemented: ${field.type}`);
    }
  }

  protected handleTextFieldChange = (event: ChangeEvent<HTMLInputElement>) => {
    const { name, value: fieldValue } = event.currentTarget;
    const { value, onChange } = this.props;

    const newValue = Object.assign({}, value, { [name]: fieldValue });

    if (onChange) {
      onChange(newValue);
    }
  };

  protected handleCheckboxChange = (event: ChangeEvent<HTMLInputElement>) => {
    const { name } = event.currentTarget;
    const checked: boolean = event.currentTarget.checked;
    const { value, onChange } = this.props;

    const newValue = Object.assign({}, value, { [name]: checked });

    if (onChange) {
      onChange(newValue);
    }
  };

  protected handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    const { onSubmit, value } = this.props;
    if (onSubmit) {
      onSubmit(value);
    }
  };

  protected isReadOnly(field: Field) {
    return !!field.readOnly;
  }
}

export default class SchemaForm extends BaseSchemaForm<{}> {}
