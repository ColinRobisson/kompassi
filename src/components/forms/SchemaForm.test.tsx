import renderer from "react-test-renderer";

import { dummyForm } from "./models";
import SchemaForm from "./SchemaForm";

describe(SchemaForm, () => {
  it("renders the dummy form", () => {
    const component = renderer.create(
      <SchemaForm
        schema={dummyForm}
        onSubmit={(values) => console.log(values)}
      />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
