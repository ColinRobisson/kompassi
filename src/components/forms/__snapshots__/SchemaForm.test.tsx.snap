// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchemaForm renders the dummy form 1`] = `
<form
  className=""
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label
      className="form-label"
      htmlFor="requiredField"
    >
      Required text field*
    </label>
    <input
      className="form-control"
      id="requiredField"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
    <small
      className="text-muted form-text"
    >
      This is the help text for the required text field which is required.
    </small>
  </div>
  <div
    className="form-group"
  >
    <label
      className="form-label"
      htmlFor="optionalField"
    >
      Optional text field
    </label>
    <input
      className="form-control"
      id="optionalField"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="pb-3"
  />
  <div
    className="form-group"
  >
    <div
      className="form-check"
    >
      <input
        className="form-check-input"
        id="requiredCheckbox"
        onChange={[Function]}
        required={true}
        type="checkbox"
        value=""
      />
      <label
        className="form-check-label"
        htmlFor="requiredCheckbox"
      >
        Required checkbox*
      </label>
      <small
        className="text-muted mb-3 form-text"
      >
        This checkbox is required. You need to check it.
      </small>
    </div>
     
  </div>
  <div
    className="form-group"
  >
    <div
      className="form-check"
    >
      <input
        className="form-check-input"
        id="optionalCheckbox"
        onChange={[Function]}
        required={false}
        type="checkbox"
        value=""
      />
      <label
        className="form-check-label"
        htmlFor="optionalCheckbox"
      >
        Optional checkbox
      </label>
      <small
        className="text-muted mb-3 form-text"
      >
        This checkbox is not required. You don't need to check it.
      </small>
    </div>
     
  </div>
  <button
    className="btn btn-primary"
    disabled={false}
    type="submit"
  >
    Submit
  </button>
</form>
`;
